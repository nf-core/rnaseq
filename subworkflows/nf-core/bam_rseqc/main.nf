//
// Run RSeQC modules
//

include { RSEQC_BAMSTAT            } from '../../../modules/nf-core/rseqc/bamstat/main'
include { RSEQC_INNERDISTANCE      } from '../../../modules/nf-core/rseqc/innerdistance/main'
include { RSEQC_INFEREXPERIMENT    } from '../../../modules/nf-core/rseqc/inferexperiment/main'
include { RSEQC_JUNCTIONANNOTATION } from '../../../modules/nf-core/rseqc/junctionannotation/main'
include { RSEQC_JUNCTIONSATURATION } from '../../../modules/nf-core/rseqc/junctionsaturation/main'
include { RSEQC_READDISTRIBUTION   } from '../../../modules/nf-core/rseqc/readdistribution/main'
include { RSEQC_READDUPLICATION    } from '../../../modules/nf-core/rseqc/readduplication/main'
include { RSEQC_TIN                } from '../../../modules/nf-core/rseqc/tin/main'

workflow BAM_RSEQC {
    take:
    bam_bai       // channel: [ val(meta), [ bam, bai ] ]
    bed           // channel: [ genome.bed ]
    rseqc_modules //    list: rseqc modules to run

    main:

    bam = bam_bai.map{ [ it[0], it[1] ] }

    versions = Channel.empty()

    //
    // Run RSeQC bam_stat.py
    //
    bamstat_txt = Channel.empty()

    if ('bam_stat' in rseqc_modules) {
        RSEQC_BAMSTAT(bam)
        bamstat_txt = RSEQC_BAMSTAT.out.txt
        versions    = versions.mix(RSEQC_BAMSTAT.out.versions.first())
    }

    //
    // Run RSeQC inner_distance.py
    //
    innerdistance_all      = Channel.empty()
    innerdistance_distance = Channel.empty()
    innerdistance_freq     = Channel.empty()
    innerdistance_mean     = Channel.empty()
    innerdistance_pdf      = Channel.empty()
    innerdistance_rscript  = Channel.empty()

    if ('inner_distance' in rseqc_modules) {
        RSEQC_INNERDISTANCE(bam, bed)
        innerdistance_distance = RSEQC_INNERDISTANCE.out.distance
        innerdistance_freq     = RSEQC_INNERDISTANCE.out.freq
        innerdistance_mean     = RSEQC_INNERDISTANCE.out.mean
        innerdistance_pdf      = RSEQC_INNERDISTANCE.out.pdf
        innerdistance_rscript  = RSEQC_INNERDISTANCE.out.rscript
        innerdistance_all      = innerdistance_distance.mix(innerdistance_freq, innerdistance_mean, innerdistance_pdf, innerdistance_rscript)
        versions               = versions.mix(RSEQC_INNERDISTANCE.out.versions.first())
    }

    //
    // Run RSeQC infer_experiment.py
    //
    inferexperiment_txt = Channel.empty()
    if ('infer_experiment' in rseqc_modules) {
        RSEQC_INFEREXPERIMENT(bam, bed)
        inferexperiment_txt = RSEQC_INFEREXPERIMENT.out.txt
        versions            = versions.mix(RSEQC_INFEREXPERIMENT.out.versions.first())
    }

    //
    // Run RSeQC junction_annotation.py
    //
    junctionannotation_all          = Channel.empty()
    junctionannotation_bed          = Channel.empty()
    junctionannotation_interact_bed = Channel.empty()
    junctionannotation_xls          = Channel.empty()
    junctionannotation_pdf          = Channel.empty()
    junctionannotation_events_pdf   = Channel.empty()
    junctionannotation_rscript      = Channel.empty()
    junctionannotation_log          = Channel.empty()

    if ('junction_annotation' in rseqc_modules) {
        RSEQC_JUNCTIONANNOTATION(bam, bed)
        junctionannotation_bed          = RSEQC_JUNCTIONANNOTATION.out.bed
        junctionannotation_interact_bed = RSEQC_JUNCTIONANNOTATION.out.interact_bed
        junctionannotation_xls          = RSEQC_JUNCTIONANNOTATION.out.xls
        junctionannotation_pdf          = RSEQC_JUNCTIONANNOTATION.out.pdf
        junctionannotation_events_pdf   = RSEQC_JUNCTIONANNOTATION.out.events_pdf
        junctionannotation_rscript      = RSEQC_JUNCTIONANNOTATION.out.rscript
        junctionannotation_log          = RSEQC_JUNCTIONANNOTATION.out.log
        junctionannotation_all          = junctionannotation_bed.mix(junctionannotation_interact_bed, junctionannotation_xls, junctionannotation_pdf, junctionannotation_events_pdf, junctionannotation_rscript, junctionannotation_log)
        versions                        = versions.mix(RSEQC_JUNCTIONANNOTATION.out.versions.first())
    }

    //
    // Run RSeQC junction_saturation.py
    //
    junctionsaturation_all     = Channel.empty()
    junctionsaturation_pdf     = Channel.empty()
    junctionsaturation_rscript = Channel.empty()

    if ('junction_saturation' in rseqc_modules) {
        RSEQC_JUNCTIONSATURATION(bam, bed)
        junctionsaturation_pdf     = RSEQC_JUNCTIONSATURATION.out.pdf
        junctionsaturation_rscript = RSEQC_JUNCTIONSATURATION.out.rscript
        junctionsaturation_all     = junctionsaturation_pdf.mix(junctionsaturation_rscript)
        versions                   = versions.mix(RSEQC_JUNCTIONSATURATION.out.versions.first())
    }

    //
    // Run RSeQC read_distribution.py
    //
    readdistribution_txt = Channel.empty()

    if ('read_distribution' in rseqc_modules) {
        RSEQC_READDISTRIBUTION(bam, bed)
        readdistribution_txt = RSEQC_READDISTRIBUTION.out.txt
        versions            = versions.mix(RSEQC_READDISTRIBUTION.out.versions.first())
    }

    //
    // Run RSeQC read_duplication.py
    //
    readduplication_all     = Channel.empty()
    readduplication_seq_xls = Channel.empty()
    readduplication_pos_xls = Channel.empty()
    readduplication_pdf     = Channel.empty()
    readduplication_rscript = Channel.empty()

    if ('read_duplication' in rseqc_modules) {
        RSEQC_READDUPLICATION(bam )
        readduplication_seq_xls = RSEQC_READDUPLICATION.out.seq_xls
        readduplication_pos_xls = RSEQC_READDUPLICATION.out.pos_xls
        readduplication_pdf     = RSEQC_READDUPLICATION.out.pdf
        readduplication_rscript = RSEQC_READDUPLICATION.out.rscript
        readduplication_all     = readduplication_seq_xls.mix(readduplication_pos_xls, readduplication_pdf, readduplication_rscript)
        versions                = versions.mix(RSEQC_READDUPLICATION.out.versions.first())
    }

    //
    // Run RSeQC tin.py
    //
    tin_txt = Channel.empty()

    if ('tin' in rseqc_modules) {
        RSEQC_TIN(bam_bai, bed)
        tin_txt      = RSEQC_TIN.out.txt
        versions    = versions.mix(RSEQC_TIN.out.versions.first())
    }

    emit:
    bamstat_txt                     // channel: [ val(meta), txt ]

    innerdistance_all               // channel: [ val(meta), {txt, pdf, r} ]
    innerdistance_distance          // channel: [ val(meta), txt ]
    innerdistance_freq              // channel: [ val(meta), txt ]
    innerdistance_mean              // channel: [ val(meta), txt ]
    innerdistance_pdf               // channel: [ val(meta), pdf ]
    innerdistance_rscript           // channel: [ val(meta), r   ]

    inferexperiment_txt             // channel: [ val(meta), txt ]

    junctionannotation_all          // channel: [ val(meta), {bed, xls, pdf, r, log} ]
    junctionannotation_bed          // channel: [ val(meta), bed ]
    junctionannotation_interact_bed // channel: [ val(meta), bed ]
    junctionannotation_xls          // channel: [ val(meta), xls ]
    junctionannotation_pdf          // channel: [ val(meta), pdf ]
    junctionannotation_events_pdf   // channel: [ val(meta), pdf ]
    junctionannotation_rscript      // channel: [ val(meta), r   ]
    junctionannotation_log          // channel: [ val(meta), log ]

    junctionsaturation_all          // channel: [ val(meta), {pdf, r} ]
    junctionsaturation_pdf          // channel: [ val(meta), pdf ]
    junctionsaturation_rscript      // channel: [ val(meta), r   ]

    readdistribution_txt            // channel: [ val(meta), txt ]

    readduplication_all             // channel: [ val(meta), {xls, pdf, r} ]
    readduplication_seq_xls         // channel: [ val(meta), xls ]
    readduplication_pos_xls         // channel: [ val(meta), xls ]
    readduplication_pdf             // channel: [ val(meta), pdf ]
    readduplication_rscript         // channel: [ val(meta), r   ]

    tin_txt                         // channel: [ val(meta), txt ]

    versions                        // channel: [ versions.yml ]
}
