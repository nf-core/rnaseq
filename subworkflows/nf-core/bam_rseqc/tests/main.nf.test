nextflow_workflow {

    name "Test Workflow BAM_RSEQC"
    script "../main.nf"
    workflow "BAM_RSEQC"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/bam_rseqc"
    tag "bam_rseqc"
    tag "rseqc"
    tag "rseqc/bamstat"
    tag "rseqc/inferexperiment"
    tag "rseqc/innerdistance"
    tag "rseqc/junctionannotation"
    tag "rseqc/junctionsaturation"
    tag "rseqc/readdistribution"
    tag "rseqc/readduplication"
    tag "rseqc/tin"

    test("sarscov2 paired-end [bam]") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed12', checkIfExists: true)
                ])
                input[2] = ['bam_stat', 'inner_distance', 'infer_experiment', 'junction_annotation', 'junction_saturation', 'read_distribution', 'read_duplication', 'tin']
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out.bamstat_txt).match("bamstat_txt")},

                { assert snapshot(workflow.out.innerdistance_all.findAll { it[1].endsWith('.pdf') == false }).match("innerdistance_all")},
                { assert workflow.out.innerdistance_all.any { it[1].endsWith('.pdf') && file(it[1]).exists() } },

                { assert snapshot(workflow.out.inferexperiment_txt).match("inferexperiment_txt")},

                { assert snapshot(workflow.out.junctionannotation_all.findAll {
                    it[1].endsWith('.xls') == false &&
                    it[1].endsWith('.r') == false }).match("junctionannotation_all")},

                { assert snapshot(workflow.out.junctionsaturation_all.findAll { it[1].endsWith('.pdf') == false }).match("junctionsaturation_all")},
                { assert workflow.out.junctionsaturation_all.any { it[1].endsWith('.pdf') && file(it[1]).exists() } },

                { assert snapshot(workflow.out.readdistribution_txt).match("readdistribution_txt")},

                { assert snapshot(workflow.out.readduplication_all.findAll { it[1].endsWith('.pdf') == false }).match("readduplication_all")},
                { assert workflow.out.readduplication_all.any { it[1].endsWith('.pdf') && file(it[1]).exists() } },

                { assert snapshot(workflow.out.tin_txt).match("tin_txt")},
                { assert snapshot(workflow.out.versions).match("versions")},
            )
        }
    }

    test("sarscov2 paired-end [bam] no modules") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed12', checkIfExists: true)
                ])
                input[2] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bamstat_txt.size() == 0 },
                { assert workflow.out.innerdistance_all.size() == 0 },
                { assert workflow.out.inferexperiment_txt.size() == 0 },
                { assert workflow.out.junctionannotation_all.size() == 0 },
                { assert workflow.out.junctionsaturation_all.size() == 0 },
                { assert workflow.out.readdistribution_txt.size() == 0 },
                { assert workflow.out.readduplication_all.size() == 0 },
                { assert workflow.out.tin_txt.size() == 0 },
                { assert workflow.out.versions.size() == 0 },
            )
        }
    }

}
