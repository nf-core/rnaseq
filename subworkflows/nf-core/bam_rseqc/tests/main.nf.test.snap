{
    "bamstat_txt": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.bam_stat.txt:md5,2675857864c1d1139b2a19d25dc36b09"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-12-08T14:37:45.671792186"
    },
    "readduplication_all": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.DupRate_plot.r:md5,3c0325095cee4835b921e57d61c23dca"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.pos.DupRate.xls:md5,a859bc2031d46bf1cc4336205847caa3"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.seq.DupRate.xls:md5,ee8783399eec5a18522a6f08bece338b"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-12-08T14:37:45.927253593"
    },
    "tin_txt": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.paired_end.sorted.summary.txt:md5,9d98447e178b89a89f6f5aba7a772fe6"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-06T15:29:37.545261"
    },
    "junctionsaturation_all": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.junctionSaturation_plot.r:md5,caa6e63dcb477aabb169882b2f30dadd"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-12-08T14:37:45.825157699"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,60dc1afce7fecb7270e998a00ee393e2",
                "versions.yml:md5,9bc3caee6aeb54f23f5296b499546515",
                "versions.yml:md5,a25161998ca60d5ce4e9a86abbf1bcb8",
                "versions.yml:md5,c175b92f50b5be38a8808cbf7ac7452e",
                "versions.yml:md5,ca7e95cf7ff7cff5d052b890729f7ead",
                "versions.yml:md5,d03beea3c68934c3f3623a005c1424d2",
                "versions.yml:md5,f2f3ecd549045f7595245f904f5d9414",
                "versions.yml:md5,fd16f1098b9c285f3ea7bd3daf4e8f10"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-12-14T14:58:10.874440924"
    },
    "readdistribution_txt": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.read_distribution.txt:md5,56893fdc0809d968629a363551a1655f"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-12-08T14:37:45.841855468"
    },
    "innerdistance_all": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.inner_distance.txt:md5,a1acc9def0f64a5500d4c4cb47cbe32b"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.inner_distance_freq.txt:md5,3fc037501f5899b5da009c8ce02fc25e"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.inner_distance_mean.txt:md5,58398b7d5a29a5e564f9e3c50b55996c"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.inner_distance_plot.r:md5,5859fbd5b42046d47e8b9aa85077f4ea"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-12-08T14:37:45.734509133"
    },
    "inferexperiment_txt": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.infer_experiment.txt:md5,f9d0bfc239df637cd8aeda40ade3c59a"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-06T15:29:37.537104"
    },
    "junctionannotation_all": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.junction_annotation.log:md5,d75e0f5d62fada8aa9449991b209554c"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-12-14T17:42:43.31783911"
    }
}